src/av_stubs.c

```diff
@@ -0,0 +1,717 @@
+  for(; frame_kind == undefined;) {
+    if( ! av->end_of_file) {
+
+      if(packet->size <= 0) {
+	// read frames from the file
+	caml_release_runtime_system();
+	int ret = av_read_frame(av->format, packet);
+	caml_acquire_runtime_system();
+
+	if(ret < 0) {
+	  packet->data = NULL;
+	  packet->size = 0;
+	  av->end_of_file = 1;
+	}
+	else if (packet->stream_index != stream_index) {
+	  av_packet_unref(packet);
+	  continue;
+	}
+      }
+    }
+    frame_kind = decode_packet(av, stream);
+  }
+  
+  if(frame_kind == end_of_file) {
+    return Val_int(END_OF_FILE_TAG);
+  }
+  else {
+    value ans = caml_alloc_small(1, media_tag);
+
+    Field(ans, 0) = stream->frame;
+    return ans;
+  }
+}
+
+CAMLprim value ocaml_av_read_audio(value _av)
+{
+  CAMLparam1(_av);
+  CAMLlocal1(ans);
+  av_t * av = Av_val(_av);
+  
+  if( ! av->audio_stream) open_audio_stream(av);
+  
+  ans = read_stream_frame(av, av->audio_stream, AUDIO_TAG);
+
+  CAMLreturn(ans);
+}
+
+CAMLprim value ocaml_av_read_video(value _av)
+{
+  CAMLparam1(_av);
+  CAMLlocal1(ans);
+  av_t * av = Av_val(_av);
+
+  if( ! av->video_stream) open_video_stream(av);
+
+  ans = read_stream_frame(av, av->video_stream, VIDEO_TAG);
+
+  CAMLreturn(ans);
+}
+
+CAMLprim value ocaml_av_read_subtitle(value _av)
+{
+  CAMLparam1(_av);
+  CAMLlocal1(ans);
+  av_t * av = Av_val(_av);
+
+  if( ! av->subtitle_stream) open_subtitle_stream(av);
+
+  ans = read_stream_frame(av, av->subtitle_stream, SUBTITLE_TAG);
+
+  if(ans != Val_int(END_OF_FILE_TAG)) {
+    Field(ans, 0) = _av;
+  }
+
+  CAMLreturn(ans);
+}
+
+CAMLprim value ocaml_av_read(value _av)
+{
+  CAMLparam1(_av);
+  CAMLlocal1(ans);
+  av_t * av = Av_val(_av);
+  AVPacket * packet = &av->packet;
+  stream_t ** streams = av->streams;
+  int nb_streams = av->format->nb_streams;
+  stream_t * stream = NULL;
+  frame_kind frame_kind = undefined;
+  
+  for(; frame_kind == undefined;) {
+    if( ! av->end_of_file) {
+  
+      if(packet->size <= 0) {
+	// read frames from the file
+	caml_release_runtime_system();
+	int ret = av_read_frame(av->format, packet);
+	caml_acquire_runtime_system();
+
+	if(ret < 0) {
+	  packet->data = NULL;
+	  packet->size = 0;
+	  av->end_of_file = 1;
+	  continue;
+	}
+      }
+  
+      if((stream = streams[packet->stream_index]) == NULL) {
+  
+	if(av->selected_streams) {
+	  av_packet_unref(packet);
+	  continue;
+	}
+	else {
+	  stream = open_stream_index(av, packet->stream_index);
+	}
+      }
+    }
+    else {
+      // If the end of file is reached, iteration on the streams to find one to flush
+      int i = 0;
+      for(; i < nb_streams; i++) {
+	if((stream = streams[i]) && stream->got_frame) break;
+      }
+  
+      if(i == nb_streams) {
+	frame_kind = end_of_file;
+	break;
+      }
+    }
+  
+    frame_kind = decode_packet(av, stream);
+  }
+  
+  if(frame_kind == end_of_file) {
+    ans = Val_int(END_OF_FILE_TAG);
+  }
+  else {
+    int media_tag = AUDIO_MEDIA_TAG;
+    value data = stream->frame;
```

Comment:
The stream->frame value being global root registered, I thought it was enough.
I will do it.